File: boats_to_save_people/boats_to_save_people.py
from typing import List

def get_number_of_rescue_boats(arr, limit):
    arr_s = sorted(arr)
    start = 0
    end = len(arr_s)-1
    boats = 0    
    while start <= end:        
        people = []
        if start != end and arr_s[start] + arr_s[end] <= limit:
            people = [arr_s[start], arr_s[end]]
            start += 1
            end -=1
            boats += 1
        else:
            people = [arr_s[end]]
            end -=1
            boats += 1
        print("{} will go with boat {}".format(people, boats))
    return boats


for arr, limit in [ [[3, 2, 3, 1], 4], [[1,2,3,4], 4], [[3, 3, 3, 2], 4]]:
    boats = get_number_of_rescue_boats(arr, limit)
    print("{}:{} => boats: {}".format(arr, limit, boats))

File: boats_to_save_people/boats_to_save_people.cpp
#include <vector>
#include <algorithm>
#include <string>
#include <boost/format.hpp>
#include <iostream>
#include "../utility/vectors.hpp"

static int get_number_of_rescue_boats(std::vector<int>& vec, int limit)
{    
    int boats = 0;
    int left = 0;
    int right = vec.size()-1;
    std::sort(begin(vec), end(vec));
    while(left <= right)
    {
        std::string people = "";
        print_vector(begin(vec)+left, begin(vec)+right+1);
        if ((left != right) && (vec[left] + vec[right]) <= limit)
        {            
            people = (boost::format("[%d, %d]") % vec[left] % vec[right]).str();            
            left += 1;
            right -= 1;
            boats += 1;
        }
        else
        {            
            people = (boost::format("[%d]") % vec[right]).str();
            right -= 1;
            boats += 1;
        }
        
        std::cout<<people<<" go with boat "<<boats<<"\n";
    }

    return boats;
}

int main()
{
    std::vector v{1, 2, 3, 3};
    auto boats = get_number_of_rescue_boats(v, 4);
    std::cout<<"Will need "<<boats<<" boats\n";
    return 0;
}